Added information about using UV (#12)
alan-turing-institute/hpc-landscape: tips-and-tricks/uv_install.md

# UV Install on HPC

UV is a new Python package manager which offers a few helpful features.

> [Website is here](https://docs.astral.sh/uv/)

Reviewer: ```suggestion
[Website is here](https://docs.astral.sh/uv/).
```
Just adding in the full stop.

This short guide doesn't explain how to use UV generally but just how to get UV to work out of a directory, which is what you would want if you wanted the Python version stored in the project space.

> 

Reviewer: Could we add a line here to mention that root access isn't needed? I think this may be a key phrase people may be looking out for.

```suggestion

This approach is likely to be especially helpful on HPC systems, since no root access, nor an existing Python installation, is required to install it this way.

```

## Installation

To install UV, run the following commands.

> It should be done in the project directory if you want all the users on the project to access it.

Reviewer: Brilliant that splitting sentences with newlines. I usually expect to have to request this change 😄

> `cd /go/to/project/directory`
> 
> ```bash
> mkdir uv_install

Reviewer: Is there a reason to separate the `cd` out from the rest of the commands here? If not, I'd suggest we pull it inside the code block:

````suggestion
```bash
cd /go/to/project/directory

mkdir uv_install
````

mkdir uv_install/bin
mkdir uv_install/pythons
mkdir uv_install/python_bins
mkdir uv_install/cache
mkdir uv_install/tools

> curl -LsSf https://astral.sh/uv/install.sh | env UV_INSTALL_DIR="${PWD}/uv_install/bin" INSTALLER_NO_MODIFY_PATH=1 sh

Reviewer: This is quite a long line, but only just overspills GitHub's standard display width. Do you think it would make sense to split it? I'm honestly not sure myself.

```suggestion
curl -LsSf https://astral.sh/uv/install.sh | \
    env UV_INSTALL_DIR="${PWD}/uv_install/bin" INSTALLER_NO_MODIFY_PATH=1 sh
```

```

This only needs to be done once per project and doesn't require `conda` or any module.

## Setting Paths

Do this all the following times you want to use UV.
It needs to be done in the same directory in which you installed it.

```bash
export PATH="${PWD}/uv_install/bin:$PATH"
export UV_CACHE_DIR="${PWD}/uv_install/cache"
export UV_PYTHON_INSTALL_DIR="${PWD}/uv_install/pythons"
export UV_PYTHON_BIN_DIR="${PWD}/uv_install/python_bins"
export UV_TOOL_DIR="${PWD}/uv_install/tools"
```

## Using UV

After it is set up, it is possible to use UV.
One thing you should do to check is run `uv python dir` to check that it points to the `uv_install` folder.
This checks if the Python binaries are being saved to the correct place and `uv python list` can be used to see if the Python version you want is installed.

> If the Python version you want isn't there, `uv python install ` can install the version number you want. 

Reviewer: ```suggestion
If the Python version you want isn't there, `uv python install` can install the version number you want. 
```
Just removing an extraneous space.

Then, it is possible to use UV as usual.
The help files can be accessed with `uv help`.

> To make an envrioment navidate to the python proejct and use `uv venv`.

Reviewer: ```suggestion
To make an environment navigate to the python project and use `uv venv`.
```
Just some small typos.

The environment can be updated by activating the environment and using `uv sync`

## SBATCH and UV

Once UV has set up the environment, it can activated with

> `source /path/to/python/project/.venv/bin/activate`

Reviewer: Would a code block make sense here?

````suggestion
```bash
source /path/to/python/project/.venv/bin/activate`
```
````

> `source /path/to/python/project/.venv/bin/activate`

Reviewer: This part about using UV with SBATCH is a really nice addition; not only is it especially relevant for HPC, it also highlights that you can use the resulting environments just like a standard virtual environment. Really nice.
